### PR TITLE
backoff exponentially if AWS throttles (resolves #1837)

### LINKED ISSUE #1837: testAutoScaledCluster sometimes is throttled to death
Example of the error trace taken from http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/1360/testReport/junit/src.toil.test.provisioners.aws.awsProvisionerTest/AWSRestartTest/testAutoScaledCluster/
```
INFO:toil.test:Setting up toil.test.provisioners.aws.awsProvisionerTest.AWSRestartTest.testAutoScaledCluster ...
INFO:toil.lib.bioio:Root logger is at level 'INFO', 'toil' logger at level 'INFO'.
INFO:toil.utils.toilLaunchCluster:Using aws provisioner.
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/venv/bin/toil", line 9, in <module>
    load_entry_point('toil', 'console_scripts', 'toil')()
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/utils/toilMain.py", line 33, in main
    module.main()
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/utils/toilLaunchCluster.py", line 93, in main
    vpcSubnet=config.vpcSubnet)
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/provisioners/aws/awsProvisioner.py", line 114, in launchCluster
    profileARN = self._getProfileARN(ctx)
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/provisioners/aws/awsProvisioner.py", line 792, in _getProfileARN
    profile = ctx.iam.create_instance_profile(iamRoleName)
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/venv/local/lib/python2.7/site-packages/boto/iam/connection.py", line 1103, in create_instance_profile
    return self.get_response('CreateInstanceProfile', params)
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/venv/local/lib/python2.7/site-packages/boto/iam/connection.py", line 102, in get_response
    raise self.ResponseError(response.status, response.reason, body)
boto.exception.BotoServerError: BotoServerError: 400 Bad Request
<ErrorResponse xmlns="https://iam.amazonaws.com/doc/2010-05-08/">
  <Error>
    <Type>Sender</Type>
    <Code>Throttling</Code>
    <Message>Rate exceeded</Message>
  </Error>
  <RequestId>6e497298-7924-11e7-86b9-d7d1cdd81d37</RequestId>
</ErrorResponse>
```

https://forums.aws.amazon.com/thread.jspa?messageID=406788&#406788 says that the recommended response is a truncated exponential back off.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-829)
┆Issue Number: TOIL-829

### PR DESCRIPTION
**Pull request by jessebrennan**
This uses a truncated exponential back-off which was recommended [here](https://forums.aws.amazon.com/thread.jspa?messageID=406788&#406788).

### CODE REVIEW

**Comment by jessebrennan** [comment]
@joelarmstrong could you review?

**Comment by ejacox** [comment]
Jenkins, test this please

File: src/toil/provisioners/aws/awsProvisioner.py [hunk 1]
```diff
@@ -781,34 +781,50 @@ def groupNotFound(e):
...
-                return profile_arn
-            else:
-                ctx.iam.remove_role_from_instance_profile(iamRoleName,
-                                                          profile.roles.member.role_name)
-        for attempt in retry(predicate=addRoleErrors):
+        def throttleError(e):
+            return e.status == 400 and e.error_code == 'Throttling'
+
+        def truncExpBackoff():
+            # as recommended here https://tinyurl.com/ycf6srh6
```
Comment (mr-c) [inline src/toil/provisioners/aws/awsProvisioner.py]: I'd use the full URL https://forums.aws.amazon.com/thread.jspa?messageID=406788#406788

File: src/toil/provisioners/aws/awsProvisioner.py [hunk 2]
```diff
@@ -781,34 +781,50 @@ def groupNotFound(e):
...
-        elif len(profile.roles) == 1:
-            # this should be profile.roles[0].role_name
-            if profile.roles.member.role_name == iamRoleName:
-                return profile_arn
-            else:
-                ctx.iam.remove_role_from_instance_profile(iamRoleName,
-                                                          profile.roles.member.role_name)
-        for attempt in retry(predicate=addRoleErrors):
+        def throttleError(e):
+            return e.status == 400 and e.error_code == 'Throttling'
```
Comment (jessebrennan) [inline src/toil/provisioners/aws/awsProvisioner.py]: Duh! good point

File: src/toil/provisioners/aws/awsProvisioner.py (hunk 1 above)
Comment (jessebrennan) [inline src/toil/provisioners/aws/awsProvisioner.py]: Yeah, shrinking wasn't really necessary and is less clear.

**Pull request merged**